### PR TITLE
[TBAA] Handle bitfields when generating !tbaa.struct metadata.

### LINKED ISSUE #82586: [TBAA][DSE] bitfield store is incorrectly optimized out
Reproducer: https://godbolt.org/z/jKrGW5jEh
```
#include <stdint.h>

struct S1 {
   signed f0 : 9;
   unsigned f1 : 2;
   int8_t  f2;
};

struct S1 g;

struct S1 func(struct S1* x) {
    x->f2 = 3;
    *x = g;
    return *x;
}
```
`x->f2 = 3` is incorrectly eliminated by `DSEPass`.
This miscompilation is caused by recent commit https://github.com/llvm/llvm-project/pull/81313.

### PR DESCRIPTION
**Pull request by fhahn**
At the moment, clang generates what I believe are incorrect !tbaa.struct fields for named bitfields. At the moment, the base type size is used for named bifields (e.g. sizeof(int)) instead of the bifield width per field. This results in overalpping fields in !tbaa.struct metadata.

This causes incorrect results when extracting individual copied fields from !tbaa.struct as in added in dc85719d5.

This patch fixes that by skipping by combining adjacent named bitfields in fields with correct size.

If this understanding is correct, I can also extend the verifier to check that !tbaa.struct fields aren't overlapping.

Fixes https://github.com/llvm/llvm-project/issues/82586

### CODE REVIEW

**Comment by efriedma-quic** [comment]
This seems like it messes up the metadata in a different way: we have no representation of the bitfield at all, so it looks like it's padding.

I think we want to treat multiple adjacent bitfields as a single "field".  So NamedBitfields from the testcase would have three fields in the LLVM TBAA data: one "field" containing both bitfields, followed by a field for the char, followed by a field for the double.  You should be able to use CGBitFieldInfo for this, I think?

**Comment by fhahn** [comment]
> This seems like it messes up the metadata in a different way: we have no representation of the bitfield at all, so it looks like it's padding.
> 
> I think we want to treat multiple adjacent bitfields as a single "field". So NamedBitfields from the testcase would have three fields in the LLVM TBAA data: one "field" containing both bitfields, followed by a field for the char, followed by a field for the double. You should be able to use CGBitFieldInfo for this, I think?

Thanks, updated the patch as suggested.

File: clang/lib/CodeGen/CodeGenTBAA.cpp [hunk 1]
```diff
@@ -294,18 +297,49 @@ CodeGenTBAA::CollectFields(uint64_t BaseOffset,
...
+
+      uint64_t Offset =
+          BaseOffset + Layout.getFieldOffset(idx) / Context.getCharWidth();
       QualType FieldQTy = i->getType();
+      // Create a single field for consecutive named bitfields using char as
+      // base type.
+      if ((*i)->isBitField() && !(*i)->isUnnamedBitfield()) {
+        unsigned CurrentBitFieldSize = 0;
+        unsigned CurrentBitFieldOffset = CGRL.getBitFieldInfo(*i).Offset;
+        while (i != e && (*i)->isBitField() && !(*i)->isUnnamedBitfield()) {
```
[the remark this answers is not in the document]
Comment (fhahn) [reply clang/lib/CodeGen/CodeGenTBAA.cpp]: Yes, test showing that added in 54cff50791dec977feb0badb74919d97dff5b859, thanks!

File: clang/lib/CodeGen/CodeGenTBAA.cpp [hunk 2]
```diff
@@ -294,18 +297,49 @@ CodeGenTBAA::CollectFields(uint64_t BaseOffset,
...
+          const CGBitFieldInfo &Info = CGRL.getBitFieldInfo(*i);
+          if (CurrentBitFieldSize + CurrentBitFieldOffset != Info.Offset)
+            break;
+          CurrentBitFieldSize += Info.Size;
```
[the remark this answers is not in the document]
Comment (fhahn) [reply clang/lib/CodeGen/CodeGenTBAA.cpp]: Thanks, updated the patch to add a field for the first bitfield in the run using its StorageSize, and skip bitfields where the offset in the run isn't 0 (`Offset`).

**Review by efriedma-quic — COMMENTED** [review]
Whitespace is weird in a few places; otherwise looks fine.

**Comment by fhahn** [comment]
> Whitespace is weird in a few places; otherwise looks fine.

Argh yes, should be fixed in the latest version. Still need to figure out how to make sure all changes are formatted once the branch contains merges (before the format checker complains and shows the commits)

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by fhahn** [comment]
Thanks!

**Pull request merged**